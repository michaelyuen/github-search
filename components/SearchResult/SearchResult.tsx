import React from "react";
import {
  License,
  NameAndDescription,
  SearchResultContainer,
  Stars,
} from "./styles";
import { ComponentProps } from "../types";

export interface SearchResultProps extends Omit<ComponentProps, "aria-label"> {
  "aria-label"?: string;
  description: string;
  id: string;
  license: string | null;
  name: string;
  owner: string;
  stars: number;
  url: string;
}

export const SearchResult: React.FC<SearchResultProps> = ({
  ["aria-label"]: ariaLabel,
  description,
  license,
  name,
  owner,
  stars,
  url,
  ...props
}) => {
  return (
    <SearchResultContainer aria-label={ariaLabel || name} {...props}>
      <NameAndDescription>
        <h3>
          <a href={url} rel="noopener noreferrer" target="_blank">
            <span>{name}/</span>
            <span>{owner}</span>
          </a>
        </h3>
        <div dangerouslySetInnerHTML={{ __html: description }} />
      </NameAndDescription>
      <Stars>
        <p>Stars:</p>
        <p>
          <a href={`${url}/stargazers`}>{stars}</a>
        </p>
      </Stars>
      <License>
        <p>License:</p>
        {license && <p>{license}</p>}
      </License>
    </SearchResultContainer>
  );
};
