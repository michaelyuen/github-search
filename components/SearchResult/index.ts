export * from "./SearchResult";
export * from "./styles";
