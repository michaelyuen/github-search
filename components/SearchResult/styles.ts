import styled from "styled-components";
import { breakpoints } from "../../styles/breakpoints";

export const SearchResultContainer = styled.li`
  background-color: var(--card-background-color);
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius);
  list-style: none;

  @media (min-width: ${breakpoints.md}) {
    display: grid;
    grid-template-columns: 1fr minmax(auto, 150px) minmax(auto, 150px);
  }
`;

export const NameAndDescription = styled.div`
  padding: 1em;

  h3 {
    margin-top: 0;
    margin-bottom: 0.5em;
  }

  a {
    display: flex;
    flex-wrap: wrap;
    font-weight: normal;

    span:last-child {
      font-weight: bold;
    }
  }

  > div div {
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
  }
`;

export const Stars = styled.div`
  border-top: 1px solid var(--border-color);
  display: flex;
  padding: 1em;

  p {
    margin: 0;

    &:first-child {
      font-weight: bold;
      margin-right: 0.5em;
    }
  }

  @media (min-width: ${breakpoints.md}) {
    display: block;
    border-left: 1px solid var(--border-color);
    border-top: none;
    text-align: center;

    p:first-child {
      margin-bottom: 1em;
      margin-right: 0;
    }
  }
`;

export const License = styled.div`
  border-top: 1px solid var(--border-color);
  display: flex;
  padding: 1em;

  p {
    margin: 0;

    &:first-child {
      font-weight: bold;
      margin-right: 0.5em;
    }
  }

  @media (min-width: ${breakpoints.md}) {
    display: block;
    border-left: 1px solid var(--border-color);
    border-top: none;
    text-align: center;

    p:first-child {
      margin-bottom: 1em;
      margin-right: 0;
    }
  }
`;
