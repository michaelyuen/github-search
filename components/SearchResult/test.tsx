import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import { SearchResult } from "./SearchResult";

describe("SearchResult", () => {
  test("renders", () => {
    render(
      <SearchResult
        aria-label="mock aria label"
        description="mock description"
        id="0"
        license="mit"
        name="name"
        stars={0}
        url="https://google.com"
      />
    );
    expect(screen.getByLabelText("mock aria label")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { container } = render(
      <SearchResult
        aria-label="mock aria label"
        description="mock description"
        id="0"
        license="mit"
        name="name"
        stars={0}
        url="https://google.com"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("has no a11y violations", async () => {
    const { container } = render(
      <SearchResult
        aria-label="mock aria label"
        description="mock description"
        id="0"
        license="mit"
        name="name"
        stars={0}
        url="https://google.com"
      />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});
