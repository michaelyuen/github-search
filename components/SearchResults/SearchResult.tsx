import { SearchResultsContainer } from "./styles";
import { SearchResult } from "../SearchResult";

export interface SearchResultsProps {
  noResultsMessage: string;
  results?: SearchResult[];
  title: string;
}

export const SearchResults: React.FC<SearchResultsProps> = ({
  noResultsMessage,
  results,
  title,
}) => {
  const hasResults = results && results.length > 0;
  return (
    <SearchResultsContainer>
      <h3>{title}</h3>
      {!hasResults && <p>{noResultsMessage}</p>}
      {hasResults &&
        results.map((result, i) => (
          <SearchResult key={result.id} {...result} />
        ))}
    </SearchResultsContainer>
  );
};
