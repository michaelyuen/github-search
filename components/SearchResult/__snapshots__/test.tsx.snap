// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult matches snapshot 1`] = `
<div>
  <ul>
    <li
      aria-label="mock aria label"
      class="sc-bcXHqe fJGIbY"
      id="0"
    >
      <div
        class="sc-gswNZR kAtizF"
      >
        <h3>
          <a
            href="https://google.com"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span>
              name
              /
            </span>
            <span>
              owner
            </span>
          </a>
        </h3>
        <div>
          mock description
        </div>
      </div>
      <div
        class="sc-dkrFOg eTqXsG"
      >
        <p>
          Stars:
        </p>
        <p>
          <a
            href="https://google.com/stargazers"
          >
            0
          </a>
        </p>
      </div>
      <div
        class="sc-hLBbgP hWJglC"
      >
        <p>
          License:
        </p>
        <p>
          mit
        </p>
      </div>
    </li>
  </ul>
</div>
`;
